Product details page

rev2023 opened the pull request:
Description
This PR adds a product details screen. The product details screen displays all the necessary data received from the API and provides an option to add products to the cart. The add-to-cart functionality is not implemented yet.

Type of change
New feature (non-breaking change which adds functionality)

Changes
The product details screen UI has been created product details screen provider has been added to manage the state of the UI. 

Testing
Manual testing has been done on Android devices of different screen sizes to test the UI.
Manual testing has been done to ensure that the details are displayed correctly.

Screenshots
![image](https://github.com/rev2023/E-commerce-application/assets/137066791/1815b043-3cf0-4780-9687-d74d6997ce8a)
![image](https://github.com/rev2023/E-commerce-application/assets/137066791/fb09f944-7e5d-4977-b130-903c3a9f5084)

Revidu-1 committed:
Implement product details screen with provider

Revidu-1 committed:
BugFix: size not selecting when clicked

Revidu-1 committed:
Code cleanup: fix spacing and use static variables for colors

MewanWA commented on lib/main.dart:
```diff
@@ -17,10 +18,12 @@ class MyApp extends StatelessWidget {
-        ChangeNotifierProvider(create: (context) => SearchScreenProvider()), // Provide the SearchScreenProvider
-        // Add more providers if needed
+        ChangeNotifierProvider(create: (context) => SearchScreenProvider()),
+        ChangeNotifierProvider(
```
Do you need to create ProductDetailsScreenProvider at this point of the app? Seems like it can be moved to the ProductDetails page

rev2023 replied:
No, it won't work as it is used in the home screen to update variables before entering the screen

MewanWA replied:
What variables need to be updated before we enter the screen?

MewanWA commented on lib/provider/home_screen_provider.dart:
```diff
@@ -64,11 +64,11 @@ class HomeScreenProvider extends ChangeNotifier {
 
   void sortPriceDescending() {
     for (int j = 1; j < productList.length; j++) {
```
Use a productList.forEach method to implement the same logic

rev2023 replied:
Done

MewanWA commented on lib/provider/home_screen_provider.dart:
```diff
@@ -79,7 +79,7 @@ class HomeScreenProvider extends ChangeNotifier {
 
   void sortPriceAscending() {
     productList.sort((a, b) =>
```
Refrain from using single lettered variable names. Name it so its easy to understand what the variables contain

MewanWA commented on lib/provider/product_details_screen_provider.dart:
```diff
@@ -0,0 +1,68 @@
+    notifyListeners(); // Notify listeners that the product has been set
+  }
+
+  Color parseColor(String colorName) {
```
Move it to a Static class

MewanWA commented on lib/provider/product_details_screen_provider.dart:
```diff
@@ -0,0 +1,68 @@
+  void populateSizes() {
+    // Populate the sizes list
+    for (String size in product.sizes) {
+      allSizes.add(Text(
```
Widgets should not be constructed in the provider classes

rev2023 replied:
Done

MewanWA reviewed (changes requested):
Update code related to comments

Revidu-1 committed:
Code cleanup: refactor methods to be more organised

This commit makes the following changes:
Moving color parsing method to static class
Changing descending order method to use the for each method and renaming method variables for clarity.
Converting the list of sizes to widgets at UI level instead of at provider.

Revidu-1 committed:
FixBug: sort in descending order not working

MewanWA commented on lib/provider/home_screen_provider.dart:
```diff
@@ -62,20 +62,18 @@ class HomeScreenProvider extends ChangeNotifier {
+      int i = productList.indexOf(product) - 1;
 
-    notifyListeners();
+      while (i >= 0 && double.parse(productList[i].price.amount) < key) {
```
We should not be having nested loops. It is not efficient. Use the sort function to do this and also for all other sort functions in this provider. 

```import 'package:flutter/foundation.dart';

class User {
  String name;
  int age;

  User(this.name, this.age);

  @override
  toString() {
    return ('$name - $age years old \n');
  }
}

void main() {
  List<User> users = [
    User('Plumber', 10),
    User('John Doe', 50),
    User('Pipi', 4),
    User('Invisible Man', 31),
    User('Electric Man', 34)
  ];

  // Age from small to large
  users.sort((a, b) => a.age.compareTo(b.age));
  debugPrint('Age ascending order: ${users.toString()}');

  // Age from large to small
  users.sort((a, b) => b.age.compareTo(a.age));
  debugPrint('Age descending order: ${users.toString()}');
}

MewanWA commented on lib/screens/home_screen.dart:
```diff
@@ -107,7 +109,8 @@ class HomeScreen extends StatelessWidget {
                                     onTap: () {
                                       homeScreenProvider.filterNike();
-                                      homeScreenProvider.selectedSortingOption = 'Regular';
+                                      homeScreenProvider.selectedSortingOption =
```
Use enums instead of directly comparing to hardcoded strings

Revidu-1 committed:
Refactor product details screen code to pass product through navigation.

Revidu-1 committed:
Using enum comparisons instead of String comparisons for sorting

Revidu-1 committed:
Code cleanup: use only package imports and fix spacing

Revidu-1 committed:
Change UI to make it look better

rev2023 commented:
> Update code related to comments

Done
